File: manual/code/test_suites/python/python/test_spacy_vector.py
import unittest
from function_spacy_vector import SpacyVectorProcessor
from unittest.mock import patch


class TestGetVectorLengths(unittest.TestCase):

    def test_function_called(self):
        text = "hello world"
        with patch('function_spacy_vector.spacy.load') as mock_load:
            processor = SpacyVectorProcessor()
            processor.vector_lengths(text)
            mock_load.assert_called_once_with('en_core_web_sm')

    def test_get_vector_lengths_length_matches_tokens(self):
        processor = SpacyVectorProcessor()
        text = "hello world"
        hello_vector_length = 8.736436  # 请根据实际运行结果更新
        world_vector_length = 7.311293  # 请根据实际运行结果更新
        expected_lengths = f"hello: {hello_vector_length:.6f}, world: {world_vector_length:.6f}"
        result_lengths = processor.vector_lengths(text)
        self.assertEqual(expected_lengths, result_lengths)

    def test_vector_lengths_with_punctuation(self):
        processor = SpacyVectorProcessor()
        text = "hello, world!"
        lengths_str = processor.vector_lengths(text)
        self.assertEqual(len(lengths_str.split(", ")), 4)  # 确保正确计算标点符号的词向量

    def test_vector_lengths_case_sensitivity(self):
        processor = SpacyVectorProcessor()
        text1 = "Hello"
        text2 = "hello"
        lengths1_str = processor.vector_lengths(text1)
        lengths2_str = processor.vector_lengths(text2)

        # 解析出词向量长度并转换为浮点数
        lengths1 = [float(length.split(":")[1].strip()) for length in lengths1_str.split(", ")]
        lengths2 = [float(length.split(":")[1].strip()) for length in lengths2_str.split(", ")]

        self.assertEqual(len(lengths1), len(lengths2))
        self.assertNotEqual(lengths1[0], lengths2[0])  # 确保它们的值不同

    def test_output_type(self):
        processor = SpacyVectorProcessor()
        text = "hello world"
        result = processor.vector_lengths(text)
        self.assertEqual(type(result), str)  # 确保返回的是字符串


if __name__ == '__main__':
    unittest.main()

File: code/test_suites/cpp/FunctionBuildTest/tests/test_spacy_vector.cpp
#include <gtest/gtest.h>
#include "../src/function_spacy_vector.cpp"

class TestGetVectorLengths : public ::testing::Test {
protected:
    SpacyVectorProcessor processor;
};

TEST_F(TestGetVectorLengths, test_function_called) {
    std::string text = "hello world";
    std::string result = processor.vector_lengths(text);
    ASSERT_FALSE(result.empty());
}

TEST_F(TestGetVectorLengths, test_get_vector_lengths_length_matches_tokens) {
    std::string text = "hello world";
    double hello_vector_length = std::sqrt(1.0*1.0 + 2.0*2.0 + 3.0*3.0);
    double world_vector_length = std::sqrt(4.0*4.0 + 5.0*5.0 + 6.0*6.0);
    std::ostringstream oss;
    oss << "hello: " << std::fixed << std::setprecision(6) << hello_vector_length
        << ", world: " << std::fixed << std::setprecision(6) << world_vector_length;
    std::string expected_lengths = oss.str();

    std::string result = processor.vector_lengths(text);
    ASSERT_EQ(expected_lengths, result);
}

TEST_F(TestGetVectorLengths, test_vector_lengths_with_punctuation) {
    std::string text = "hello, world!";
    std::string lengths_str = processor.vector_lengths(text);
    std::vector<std::string> split_result;
    boost::split(split_result, lengths_str, boost::is_any_of(", "));

    // 移除空字符串
    split_result.erase(std::remove_if(split_result.begin(), split_result.end(), [](const std::string& s) { return s.empty(); }), split_result.end());

    ASSERT_EQ(split_result.size(), 5);
}


TEST_F(TestGetVectorLengths, test_vector_lengths_case_sensitivity) {
    std::string text1 = "Hello";
    std::string text2 = "hello";

    std::string lengths1_str = processor.vector_lengths(text1);
    std::string lengths2_str = processor.vector_lengths(text2);

    auto parse_length = [](const std::string& s) {
        size_t colon_pos = s.find(":");
        return std::stod(s.substr(colon_pos + 1));
    };

    double length1 = parse_length(lengths1_str);
    double length2 = parse_length(lengths2_str);

    ASSERT_NE(length1, length2);
}

TEST_F(TestGetVectorLengths, test_output_type) {
    std::string text = "hello world";
    std::string result = processor.vector_lengths(text);
    ASSERT_TRUE(!result.empty());
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}